src/datasets/datasets-public.v4.controller.ts

```diff
@@ -0,0 +1,245 @@
+      "It returns a list of metadata keys contained in the public datasets matching the filter provided.",
+    description:
+      "It returns a list of metadata keys contained in the public datasets matching the filter provided.<br>This endpoint still needs some work on the filter and facets specification.",
+  })
+  @ApiQuery({
+    name: "fields",
+    description:
+      "Define the filter conditions by specifying the name of values of fields requested. There is also support for a `text` search to look for strings anywhere in the dataset.",
+    required: false,
+    type: String,
+    example: {},
+  })
+  @ApiQuery({
+    name: "limits",
+    description: "Define further query parameters like skip, limit, order",
+    required: false,
+    type: String,
+    example: '{ "skip": 0, "limit": 25, "order": "creationTime:desc" }',
+  })
+  @ApiResponse({
+    status: HttpStatus.OK,
+    type: String,
+    isArray: true,
+    description: "Return metadata keys for list of public datasets selected",
+  })
+  // NOTE: This one needs to be discussed as well but it gets the metadata keys from the dataset but it doesnt do it with the nested fields. Think about it
+  async publicMetadataKeys(
+    @Query() filters: { fields?: string; limits?: string },
+  ) {
+    let fields: IDatasetFields = JSON.parse(filters.fields ?? "{}");
+
+    fields = { ...fields, isPublished: true };
+
+    const parsedFilters: IFilters<DatasetDocument, IDatasetFields> = {
+      fields: fields,
+      limits: JSON.parse(filters.limits ?? "{}"),
+    };
+
+    return this.datasetsService.metadataKeys(parsedFilters);
+  }
+
+  // GET /datasets/public/findOne
+  @Get("/findOne")
+  @ApiOperation({
+    summary: "It returns the first public dataset found.",
+    description:
+      "It returns the first public dataset of the ones that matches the filter provided. The list returned can be modified by providing a filter.",
+  })
+  @ApiQuery({
+    name: "filter",
+    description: "Database filters to apply when retrieving public dataset",
+    required: true,
+    type: String,
+    content: getSwaggerDatasetFilterContent({
+      where: true,
+      include: true,
+      fields: true,
+      limits: true,
+    }),
+  })
+  @ApiResponse({
+    status: HttpStatus.OK,
+    type: OutputDatasetDto,
+    description: "Return the datasets requested",
+  })
+  async findOnePublic(
+    @Query("filter", new FilterValidationPipe(), new IncludeValidationPipe())
+    queryFilter: string,
+  ): Promise<OutputDatasetDto | null> {
+    const parsedFilter = JSON.parse(queryFilter ?? "{}");
+
+    this.addPublicFilter(parsedFilter);
+
+    const foundDataset =
+      await this.datasetsService.findOneComplete(parsedFilter);
+
+    if (!foundDataset) {
+      // TODO: Do we want to throw here if the dataset is not found!?
+      // something like: throw new NotFoundException(`Dataset with provided filters: ${queryFilter} was not found. Please check your filter and try again`);
+    }
```

Comment:
Let's keep what has been implemented on the datasets.controller for now. Even if we need to change it later it will be easier for us at that point.